<a href="https://colab.research.google.com/github/jrakhshanda/Text-Mining/blob/main/TM_L1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# L1: Information retrieval

In this lab you will apply basic techniques from information retrieval to implement the core of a minimalistic search engine. The data for this lab consists of a collection of app descriptions scraped from the [Google Play Store](https://play.google.com/store/apps?hl=en). From this collection, your search engine should retrieve those apps whose descriptions best match a given query under the vector space model.

## Data set

The app descriptions come in the form of a compressed [JSON](https://en.wikipedia.org/wiki/JSON) file. Start by loading this file into a [Pandas](https://pandas.pydata.org) [DataFrame](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dataframe).

In [2]:
import bz2
import pandas as pd
import spacy

with bz2.open('app-descriptions.json.bz2') as source:
    df = pd.read_json(source)

In [3]:
df.head()

name                                        description
0                                           10000000  10000000 is a Dungeon Crawling Puzzle RPG Mat...
1                                    1177 Vårdguiden  I 1177 Vårdguidens app får du tillgång till 11...
2  123 Draw🎨 Toddler counting for kids Drawing games  Need counting games for kids & drawing for tod...
3  123 Kids Fun MUSIC - Kids Music Educational Games  Beautiful and simple music application for tod...
4       123 number games for kids -  Count & Tracing  A Fun and intuitive numbers game for your baby...

In Pandas, a DataFrame is a table with indexed rows and labelled columns of potentially different types. Data in a DataFrame can be accessed in various ways, including by row and by column. To give an example, the code in the next cell shows rows 200–204:

In [4]:
df[200:205]

name                                        description
200  Brick Breaker Star: Space King  Introducing the best Brick Breaker game that e...
201      Brick Classic - Brick Game  Classic Brick Game!\n\nBrick Classic is a popu...
202     Bricks Breaker - Glow Balls  Bricks Breaker - Glow Balls is a addictive and...
203            Bricks Breaker Quest  How to play\n- The ball flies to wherever you ...
204             Brothers in Arms® 3  Fight brave soldiers from around the globe on ...

As you can see, there are two labelled columns: `name` (the name of the app) and `description` (a textual description). The code in the next cell shows how to acess fields from the description column.

In [5]:
df['description'][200:205]

200    Introducing the best Brick Breaker game that e...
201    Classic Brick Game!\n\nBrick Classic is a popu...
202    Bricks Breaker - Glow Balls is a addictive and...
203    How to play\n- The ball flies to wherever you ...
204    Fight brave soldiers from around the globe on ...
Name: description, dtype: object

In [6]:
df['description'][201]

"Classic Brick Game!\n\nBrick Classic is a popular and addictive puzzle game!\n\nHow to play?\n- Simply drag the bricks to move them.\n- Create full lines on the grid vertically or horizontally to break bricks.\n\nTips:\n- Classic brick game without time limits.\n- Place the bricks in a reasonable position.\n- The more brick break, the more scores you have.\n- Bricks can't be rotated.\n\nWho's the best brick breaker? Challenge it now!!!"

## Problem 1: Preprocessing

Your first task is to implement a preprocessor for your search engine. In the vector space model, *preprocessing* refers to any kind of transformation that is applied before a text is vectorized. Here you can restrict yourself to a very simple preprocessing: tokenization, stop word removal, and lemmatization.

To implement your preprocessor, you can use [spaCy](https://spacy.io). Make sure that you read the [Linguistic annotations](https://spacy.io/usage/spacy-101#annotations) section of the spaCy&nbsp;101; that section contains all the information that you need for this problem (and more).

Implement your preprocessor by completing the skeleton code in the next cell, adding additional code as you feel necessary.

In [7]:
nlp = spacy.load("en_core_web_sm",disable=["tagger", "parser","ner"])
def preprocess(text):
    # TODO: Replace the next line with your own code.
    doc = nlp(text)
    token_words = []
    for token in doc:
      if not token.is_stop and token.is_alpha:
        token_words.append(token.lemma_)
    return token_words

In [8]:
preprocess(df['description'][201])

['Classic',
 'Brick',
 'Game',
 'Brick',
 'Classic',
 'popular',
 'addictive',
 'puzzle',
 'game',
 'play',
 'Simply',
 'drag',
 'brick',
 'Create',
 'line',
 'grid',
 'vertically',
 'horizontally',
 'break',
 'brick',
 'Tips',
 'Classic',
 'brick',
 'game',
 'time',
 'limit',
 'Place',
 'brick',
 'reasonable',
 'position',
 'brick',
 'break',
 'score',
 'Bricks',
 'rotate',
 'well',
 'brick',
 'breaker',
 'Challenge']

Your implementation should conform to the following specification:

<strong>preprocess</strong> (<em>text</em>)

> Preprocesses given text by tokenizing it, removing any stop words, replacing each remaining token with its lemma (base form), and discarding all lemmas that contain non-alphabetical characters. Returns the list of remaining lemmas (represented as strings).

**Tip:** To speed up the preprocessing, you can disable loading those spaCy components that you do not need, such as the part-of-speech tagger, parser, and named entity recognizer. See [here](https://spacy.io/usage/processing-pipelines#disabling) for more information about this.

Test your implementation by running the following cell:

In [9]:
preprocess('Apple is looking at buying U.K. startup for $1 billion')

['Apple', 'look', 'buy', 'startup', 'billion']

This should give the following output:


['Apple', 'look', 'buy', 'startup', 'billion']

## Problem 2: Vectorizing

Your next task is to vectorize the data – and more specifically, to map each app description to a tf–idf vector. For this you can use the [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html) class from [scikit-learn](https://scikit-learn.org/stable/). Make sure to specify your preprocessor from the previous problem as the `tokenizer` &ndash; not the `preprocessor`! &ndash; for the vectorizer. (In scikit-learn parlance, the `preprocessor` handles string-level preprocessing.)

In [10]:
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.neighbors import NearestNeighbors

# TODO: Replace the next line with your own code.
X = np.zeros((len(df), 1))
vectorizer = TfidfVectorizer(tokenizer = preprocess)
X = vectorizer.fit_transform(df['description'])
vocabulary = vectorizer.vocabulary_
feature_names = vectorizer.get_feature_names()

Test your implementation by running the following cell:

In [11]:
X.shape

(1614, 20663)

This should show the dimensions of the matrix `X` to be 1614 × 20669.

## Problem 3: Retrieving

To complete the search engine, your last task is to write a function that returns the most relevant app descriptions for a given query. An easy way to do solve this task is to use scikit-learn&rsquo;s [NearestNeighbors](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html) class. That class implements unsupervised nearest neighbours learning, and allows you to easily find a predefined number of app descriptions whose vector representations are closest to the query vector.

In [18]:
neigh = NearestNeighbors(n_neighbors = 10, n_jobs=-1) 
neigh.fit(X)

def search(query):
    # TODO: Replace the next line with your own code.
    x = vectorizer.transform([query])
    result = neigh.kneighbors(x, return_distance=False)
    return df['name'][result[0]]

Your implementation should conform to the following specification:

<strong>search</strong> (<em>query</em>)

> Returns the 10 app descriptions most similar (in terms of cosine similarity) to the given query as a Pandas DataFrame.

Test your implementation by running the following cell:

In [19]:
search('dodge trains')

1301                               Subway Surfers
1300                       Subway Princess Runner
1428                        Train Conductor World
998                No Humanity - The Hardest Game
1394                                   Tiny Rails
1429          Train for Animals - BabyMagica free
1168                                         Rush
1077                  Polar Flow – Sync & Analyze
1286    Strava: Track Running, Cycling & Swimming
1465              Virus War - Space Shooting Game
Name: name, dtype: object

The top hit in the list should be *Subway Surfers*.

## Problem 4: Finding terms with low/high idf

Recall that the inverse document frequency (idf) of a term is the lower the more documents from a given collection the term appears in. To get a better understanding for this concept, your next task is to write code to find out which terms have the lowest/highest idf with respect to the app descriptions.

Start by sorting the terms in increasing order of idf, breaking ties by falling back on alphabetic order.

In [21]:
# TODO: Replace the next line with your own code.
tf_idf = vectorizer.idf_.argsort()
terms = [vectorizer.get_feature_names()[i] for i in tf_idf[0:len(tf_idf)]]

Now, print the 10 terms with the lowest/highest idf. How do you explain the results?

In [22]:
print(terms[:10])

['game', 'play', 'feature', 'free', 'new', 'world', 'time', 'app', 'fun', 'use']


In [23]:
print(terms[-10:])

['chiamato', 'outward', 'chia', 'outworld', 'chew', 'overblown', 'overboard', 'overcast', 'outpost', 'flying']


The terms with low idf means these terms are very common in the document and large idf means the terms are rare in the documents.

## Problem 5: Keyword extraction

A simple method for extracting salient keywords from a document is to pick the $k$ terms with the highest tf–idf value. Your last task in this lab is to implement this method. More specifically, we ask you to implement a function `keywords` that extracts keywords from a given text.

In [ ]:
def get_keywords(text, n=10):
    # TODO: Replace the next line with your own code.
    tf_idf = vectorizer.transform([text])
    sorted_items=sort_coo(tf_idf.tocoo())
    keywords = [feature_names[idx] for idx, score in sorted_items[0:n]]
    return keywords

def sort_coo(coo_matrix):
    tuples = zip(coo_matrix.col, coo_matrix.data)
    return sorted(tuples, key=lambda x: (x[1], x[0]), reverse=True)

Your implementation should conform to the following specification:

<strong>keywords</strong> (<em>text</em>, <em>n</em> = 10)

> Returns a list with the $n$ (default value: 10) most salient keywords from the specified text, as measured by their tf–idf value relative to the collection of app descriptions.

Test your implementation by running the following cell:

In [ ]:
print(get_keywords(df['description'][1428]))

['train', 'railway', 'railroad', 'rail', 'chaos', 'crash', 'tram', 'timetable', 'railyard', 'overcast']


This should give the following output:

['train', 'railway', 'railroad', 'rail', 'chaos', 'crash', 'overcast', 'locomotive', 'timetable', 'railyard']

<div class="alert alert-info">
    Please read the General information section on the lab web page before submitting this notebook!
</div>